# Exploring eBay Car Sales Data

In this project, we are going to work with a dataset of used cars from [eBay Kleinanzeigen](https://www.ebay-kleinanzeigen.de/) , a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).
 However [DataQuest](https://www.dataquest.io/) has made a few modifications to the dataset by:
- Sampling 50,000 data points from the full dataset, to ensure the code runs quickly in their hosted environment.
- Dirtying the dataset a bit more to closely resemble what we would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with).

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In this project, we are going to clean and analyze the included used car listings.

Now let's import pandas and Numpy libraries and read our file into pandas

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Our column names are in camelcase and some of them do not clearly give an idea of what they represent.
<br> Also, some of the data in our column (like `notRepairedDamege`) are in German.

In [4]:
col_header = autos.columns
header = []
for c in col_header :
    c = c.replace('yearOfRegistration', 'registration_year')
    c = c.replace('monthOfRegistration', 'registration_month')
    c = c.replace('notRepairedDamage', 'unrepaired_damage')
    c = c.replace('dateCreated', 'ad_created')
    import re
    c = re.sub(r'(?<!^)(?=[A-Z])', '_', c).lower()
    header.append(c)
autos.columns = header
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In the above step, we edited some of the column names in our dataframe to make them clearer and fiting the description.
We also converted all our column names from camelcase to snakecase.

In [5]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [6]:
print(autos['price'].head(5))
print(autos['odometer'].head(5))

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object


Looking at the above description of our data. 
<br> We notice that there are a number of text columns where almost all of the values are the same (seller and offer_type).
<br> We also notice that the 'n_of_picture' column only has zero values.
<br> Another thing to note is that the price and the odometer columns are both numeric values stores as text. So our next step will be to clean those columns (price and odometer), and covert them to numeric type.

In [7]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(int)
autos = autos.rename({'odometer':'odometer_km'}, axis = 1)

Now let's explore the 'price' and the 'odometer_km' columns and find out whether there are values that do not belong.

In [8]:
print('There are ', autos['price'].unique().shape, 'unique values in the price column')
print('\n')
autos['price'].describe()

There are  (2357,) unique values in the price column




count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [9]:
print('There are ',autos['odometer_km'].unique().shape, 'unique values in the odometer colomn')
print('\n')
autos['odometer_km'].describe()

There are  (13,) unique values in the odometer colomn




count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Let us take a look at the five lowest values in the price column and how many time listings they appear on.

In [10]:
autos['price'].value_counts().head().sort_index()

0       1421
500      781
1000     639
1500     734
2500     643
Name: price, dtype: int64

Now let's look at the five highest values in the price column

In [11]:
autos['price'].sort_values(ascending = False).head()

39705    99999999
42221    27322222
39377    12345678
47598    12345678
27371    12345678
Name: price, dtype: int64

Now let us take a look at all the values in the odometer colomn.

In [12]:
autos['odometer_km'].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

After exploring the price and the odometer_km column, we notice that there are listings that have a price of zero (1421 instances of a price of zero to be exact).
<br> We also notice that the the five highest prices in our data are respetively 99,999,999 , 27,322,222 , 12,345,678 , 12,345,678 and 2,345,678.
<br> The probability of a seller taking the time to create a post just to give the car away for free is very slim. Even cars that do not run anymore usually sell for at least \$150
<br> Also, we do not usually see people purchasing cars worth millions of dollar on ebay.
<br> Because of those two reason, we have decided to clean our data and only include the rows where the car price is between 200 and 200000 USD.


In [13]:
autos = autos[autos['price'].between(150,200001)]

In [14]:
autos.describe(include = 'all')

date_crawled      name  seller offer_type          price  \
count                 47990     47990   47990      47990   47990.000000   
unique                46350     36975       1          1            NaN   
top     2016-03-19 17:36:18  BMW_316i  privat    Angebot            NaN   
freq                      3        75   47990      47990            NaN   
mean                    NaN       NaN     NaN        NaN    5911.193686   
std                     NaN       NaN     NaN        NaN    8326.812376   
min                     NaN       NaN     NaN        NaN     150.000000   
25%                     NaN       NaN     NaN        NaN    1271.250000   
50%                     NaN       NaN     NaN        NaN    3100.000000   
75%                     NaN       NaN     NaN        NaN    7500.000000   
max                     NaN       NaN     NaN        NaN  198000.000000   

       abtest vehicle_type  registration_year  gearbox     power_p_s  model  \
count   47990        43642       47990.000000    45836  47990.000000  45628   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    24698        12537                NaN    35754           NaN   3860   
mean      NaN          NaN        2004.759408      NaN    117.892457    NaN   
std       NaN          NaN          88.108860      NaN    200.816203    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     75.000000    NaN   
50%       NaN          NaN        2004.000000      NaN    109.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    47990.000000        47990.000000     44184       47990   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     29051       10224   
mean    125944.050844            5.810356       NaN         NaN   
std      39473.314917            3.672475       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              39230                47990         47990.0  47990.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               34675                 1865             NaN           NaN   
mean                 NaN                  NaN             0.0  50996.420629   
std                  NaN                  NaN             0.0  25731.100479   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30823.000000   
50%                  NaN                  NaN             0.0  49733.000000   
75%                  NaN                  NaN             0.0  71672.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 47990  
unique                38065  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Now, let's move on to the data column, so we can understand the date range the data covers.
<br> Of the five columns that represent date values. Some were created by the crawler, and some came from the website itself. Referring to the dta dictionary:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [15]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025359
2016-03-06    0.014086
2016-03-07    0.035945
2016-03-08    0.033257
2016-03-09    0.033049
2016-03-10    0.032361
2016-03-11    0.032673
2016-03-12    0.036883
2016-03-13    0.015732
2016-03-14    0.036653
2016-03-15    0.034236
2016-03-16    0.029464
2016-03-17    0.031527
2016-03-18    0.012878
2016-03-19    0.034695
2016-03-20    0.037820
2016-03-21    0.037320
2016-03-22    0.032882
2016-03-23    0.032319
2016-03-24    0.029402
2016-03-25    0.031423
2016-03-26    0.032215
2016-03-27    0.031111
2016-03-28    0.034924
2016-03-29    0.034049
2016-03-30    0.033778
2016-03-31    0.031840
2016-04-01    0.033715
2016-04-02    0.035632
2016-04-03    0.038612
2016-04-04    0.036528
2016-04-05    0.013086
2016-04-06    0.003167
2016-04-07    0.001375
Name: date_crawled, dtype: float64

Judging from the 'date crawled' column, our listing were crawled between March 5th 2016 and April 7th 2016

In [16]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000042
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000063
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000042
2016-02-05    0.000042
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000042
2016-02-14    0.000042
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000042
2016-02-19    0.000063
2016-02-20    0.000042
2016-02-21    0.000063
                ...   
2016-03-09    0.033132
2016-03-10    0.032069
2016-03-11    0.032986
2016-03-12    0.036716
2016-03-13    0.017108
2016-03-14    0.035278
2016-03-15    0.033965
2016-03-16    0.029965
2016-03-17    0.031194
2016-03-18    0.013565
2016-03-19    0.033549
2016-03-20    0.037904
2016-03-21 

As we can see above, the range for the date our ads were created is far longer and goes for a span of approximatly 10 months from June 11th 2015 to April 7th,2016.

In [17]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001084
2016-03-06    0.004313
2016-03-07    0.005376
2016-03-08    0.007293
2016-03-09    0.009585
2016-03-10    0.010627
2016-03-11    0.012419
2016-03-12    0.023817
2016-03-13    0.008877
2016-03-14    0.012628
2016-03-15    0.015816
2016-03-16    0.016399
2016-03-17    0.028048
2016-03-18    0.007314
2016-03-19    0.015732
2016-03-20    0.020650
2016-03-21    0.020588
2016-03-22    0.021421
2016-03-23    0.018525
2016-03-24    0.019754
2016-03-25    0.019129
2016-03-26    0.016712
2016-03-27    0.015482
2016-03-28    0.020733
2016-03-29    0.022275
2016-03-30    0.024651
2016-03-31    0.023859
2016-04-01    0.022901
2016-04-02    0.024818
2016-04-03    0.025130
2016-04-04    0.024526
2016-04-05    0.125109
2016-04-06    0.222067
2016-04-07    0.132340
Name: last_seen, dtype: float64

The range of when the crawler last saw the add is similar to the range date of when the add was first crawled, ranging from March 5th 2016 to April 7th, 2016.

Now let's take a close look at the registration year columns

In [18]:
autos['registration_year'].describe()

count    47990.000000
mean      2004.759408
std         88.108860
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

A close look at the registration_year column shows the earliest regitration year for card in our listings being the year 1000 (before cars were invented) and the latest being 9999 (a year even our great -great - great - great, ... ,great- great- great grand-children, will not get to experience). This shows that there are errors that we need to clean in this column.

First let's look at the number of listings for cars that have a registration year between 1000 and 1950.

In [19]:
auto_1 = autos[autos['registration_year'].between(1000, 1950)]

In [20]:
auto_1['registration_year'].describe()

count      23.000000
mean     1805.869565
std       307.514124
min      1000.000000
25%      1910.000000
50%      1934.000000
75%      1938.500000
max      1950.000000
Name: registration_year, dtype: float64

There are 23 listings with a registration year between 1000 and 1950.
<br> Because the mean registration year for the cars in those listings is 1934, we are going to use 1930 as our earliest valid registration year.
<br> And because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. So 2016 is going to be our latest valid registration year.
<br> Now let's remove the values that fall outside this interval.

In [21]:
autos = autos[autos['registration_year'].between(1930, 2016)]

In [22]:
autos['registration_year'].value_counts(normalize = True).sort_index()

1931    0.000022
1934    0.000043
1937    0.000087
1938    0.000022
1939    0.000022
1941    0.000043
1943    0.000022
1948    0.000022
1950    0.000022
1951    0.000043
1952    0.000022
1953    0.000022
1954    0.000043
1955    0.000043
1956    0.000087
1957    0.000043
1958    0.000087
1959    0.000130
1960    0.000455
1961    0.000130
1962    0.000087
1963    0.000173
1964    0.000260
1965    0.000369
1966    0.000477
1967    0.000564
1968    0.000564
1969    0.000412
1970    0.000781
1971    0.000564
          ...   
1987    0.001561
1988    0.002884
1989    0.003686
1990    0.007047
1991    0.007264
1992    0.007893
1993    0.008999
1994    0.013314
1995    0.025262
1996    0.028991
1997    0.041178
1998    0.050198
1999    0.062146
2000    0.066418
2001    0.056899
2002    0.053624
2003    0.058395
2004    0.058503
2005    0.063035
2006    0.057853
2007    0.049266
2008    0.047835
2009    0.045081
2010    0.034412
2011    0.035063
2012    0.028319
2013    0.017325
2014    0.0143

The registration year for the cars in our listings is distributed over several years with no sigle year being overly dominant.
<br> However, we can notice that the a sizeable portion of the cars in our listings were first registered in the period between 1998 to 2016 with every single year in that period having more than 5% of the total cars registered.

Now let's move on and explore variations accross different car brands.
<br> Let's take a look at all the diferent brands in our data and how they compare in term of popularity.

In [23]:
autos['brand'].value_counts().sort_values(ascending = False)

volkswagen        9754
bmw               5099
opel              4922
mercedes_benz     4477
audi              4015
ford              3195
renault           2161
peugeot           1379
fiat              1171
seat               843
skoda              760
nissan             707
mazda              701
smart              658
citroen            649
toyota             593
hyundai            464
sonstige_autos     437
volvo              422
mini               407
mitsubishi         377
honda              363
kia                328
alfa_romeo         309
porsche            274
suzuki             271
chevrolet          262
chrysler           163
dacia              123
daihatsu           115
jeep               106
subaru              98
land_rover          98
saab                77
jaguar              71
daewoo              68
trabant             63
rover               61
lancia              49
lada                27
Name: brand, dtype: int64

Looking at the above brand counts, we notice the Volkswagen, Bmw, Opel, Mercedes_benz, and Audi are the 5 most popular brands in our data(with Volkswagen far ahead of the pack).
<br> In the rest of our analyzis, are going to focus on the brands that have 800+ listings in our data. Those also happen to be the top 10 brands (popularity wise) in our data.

First, let's look at the average price per brands for the brands in our chosen top 10

In [24]:
brand_counts = autos['brand'].value_counts()
top10_brands = brand_counts[brand_counts > 800].index
print(top10_brands)


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [25]:
avg_price_top10brands = {}
for brand in top10_brands :
    avg_price = autos.loc[autos['brand'] == brand, 'price'].mean()
    avg_price_top10brands[brand] = int(avg_price)
    
print(avg_price_top10brands)

{'volkswagen': 5461, 'bmw': 8292, 'opel': 3034, 'ford': 3822, 'audi': 9396, 'peugeot': 3124, 'mercedes_benz': 8678, 'seat': 4448, 'fiat': 2874, 'renault': 2519}


Looking at our top 10 brands:
- Audi, Mercedes_benz and Bmw are respectively the most expennsive brands.
- Volkswagen, and Seat are around the middle when it comes to price, which might explain why one of them (Volkswagen) is so popular (might be offering very good value for quality).
- Ford, Opel, Peugeot, Fiat, and Renault are on the lower side of the price spectrum, with Renault being the least expensive brand in our top 10

In [26]:
mean_price = pd.Series(avg_price_top10brands)
pd.DataFrame(mean_price, columns = ['mean_price'])

mean_price
audi                 9396
bmw                  8292
fiat                 2874
ford                 3822
mercedes_benz        8678
opel                 3034
peugeot              3124
renault              2519
seat                 4448
volkswagen           5461

Now let's find out the average car mileage for the top 10 brands, and determine whether there is a relationship between the average price per brands that we computed earlier and the average car mileage.

In [27]:
top10b_avg_miles = {}
for brand in top10_brands:
    avg_miles = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    top10b_avg_miles[brand] = int(avg_miles)
    
print(top10b_avg_miles)

{'volkswagen': 128800, 'bmw': 132812, 'opel': 129363, 'ford': 124245, 'audi': 129281, 'peugeot': 127063, 'mercedes_benz': 131077, 'seat': 121690, 'fiat': 116810, 'renault': 128246}


In [28]:
mean_mileage = pd.Series(top10b_avg_miles).sort_values(ascending=False)
mean_prices = pd.Series(avg_price_top10brands).sort_values(ascending=False)

brand_info = pd.DataFrame(mean_mileage, columns = ['mean_miles'])
brand_info

mean_miles
bmw                132812
mercedes_benz      131077
opel               129363
audi               129281
volkswagen         128800
renault            128246
peugeot            127063
ford               124245
seat               121690
fiat               116810

In [29]:
brand_info['mean_price'] = mean_price
brand_info

mean_miles  mean_price
bmw                132812        8292
mercedes_benz      131077        8678
opel               129363        3034
audi               129281        9396
volkswagen         128800        5461
renault            128246        2519
peugeot            127063        3124
ford               124245        3822
seat               121690        4448
fiat               116810        2874

There isn't a clear relationship between the car mileage and the car price.
There is however a slight trends where the expensive cars have more mileage than the cheap one